src/fieldAnalyser/fieldAnalyser.ts

```diff
@@ -0,0 +1,138 @@
+import PlatformClient, {FieldTypes} from '@coveord/platform-client';
+import {BatchUpdateDocuments, DocumentBuilder} from '..';
+
+type fieldTypeDict = {[key in FieldTypes]?: number};
+type typeOccurence = [fieldType: string, occurence: number];
+export class FieldAnalyser {
+  private static fieldTypePrecedence = ['DOUBLE', 'STRING'];
+  private inconsistentFields: Record<string, FieldTypes[]> = {};
+
+  public constructor(private platformClient: PlatformClient) {}
+  public async createMissingFieldsFromBatch(batch: BatchUpdateDocuments) {
+    const alreadyCreatedFields = await this.listAllFieldsFromOrg();
+    const missingFieldsFromOrg = this.getMissingFieldsFromOrg(
+      batch,
+      alreadyCreatedFields
+    );
+    const fieldsToCreate = this.getFieldsToCreate(missingFieldsFromOrg);
+    return this.createFields(fieldsToCreate);
+  }
+
+  private async listAllFieldsFromOrg(): Promise<string[]> {
+    // TODO: CDX-836
+    throw new Error('Method not implemented.');
+  }
+
+  private createFields(_fields: Record<string, FieldTypes>) {
+    // TODO: CDX-835
+    throw new Error('Method not implemented.');
+  }
+
+  private getMissingFieldsFromOrg(
+    batch: BatchUpdateDocuments,
+    alreadyCreatedFields: string[]
+  ): Record<string, fieldTypeDict> {
+    const missingFieldsFromOrg: Record<string, fieldTypeDict> = {};
+
+    batch.addOrUpdate.flatMap((doc: DocumentBuilder) => {
+      const documentMetadata = Object.entries({...doc.build().metadata});
+
+      for (const [metadataKey, metadataValue] of documentMetadata) {
+        if (alreadyCreatedFields.includes(metadataKey)) {
+          continue;
+        }
+        const metadataType = this.getValue(metadataValue);
+        if (missingFieldsFromOrg[metadataKey]) {
+          // Possible metadata inconsistency
+          let fieldType = missingFieldsFromOrg[metadataKey][metadataType];
+          missingFieldsFromOrg[metadataKey][metadataType] = fieldType
+            ? ++fieldType
+            : 1;
+        } else {
+          missingFieldsFromOrg[metadataKey] = {[metadataType]: 1};
+        }
```

Comment:
Create a method for that, it'll help with readability

Reply:
Sorry but I don't agree here :) 
We need a class so we can instantiate an objet to hold some kind of memory. I know that right now it does not make sense on its own, but you will understand once the `FieldAnalyser` is added to the `Source` class.

The idea of having a class (and not a function ) is to prevent consuming all documents batches at the same time.
That way, we can have an analyser that can consume batches iteratively without having to load all the documents in memory.

Reply:
For whoever may read this in the future:
 - @y-lakhdar answer was meant to address this comment: https://github.com/coveo/push-api-client.js/pull/91#pullrequestreview-890333290
 - The usage of a class was indeed appropriate: The job of reading documents & putting them in a standard object is done by the `DocBuilder` classes. Because we cannot store all documents in memory, we need to keep the result of the ongoing analysis (i.e., the list of fields already existing, the list of missing fields and their inferred types)